### ISSUE TITLE
SIGSEGV after restore

### ISSUE THREAD

**Issue by nezed** [VictoriaMetrics/VictoriaMetrics#468]
Hi!
I'm trying to restore data from single-node to vmstorage but when i starting vmstorage with new data it always crashes with error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x7c5b29]
goroutine 1 [running]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*indexBlockCache).MustClose(0x0, 0xc000052c00)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/part.go:206 +0x29
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*part).MustClose(0xc00023a090)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/part.go:148 +0x8c
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.newPart(0xc00011f1d0, 0xc00008af80, 0x7c, 0x28e8e, 0xb56520, 0xc00007de90, 0xb565a0, 0xc00035b7a0, 0xb565a0, 0xc00035b7c0, ...)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/part.go:124 +0x359
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.openFilePart(0xc00008af80, 0x7c, 0xee16e0, 0xa919e5, 0x1)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/part.go:96 +0x7a8
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.openParts(0xc0000f8a40, 0x3b, 0xc0000f8a80, 0x39, 0xc0000f8a40, 0x3b, 0x0, 0x0, 0x0, 0x0, ...)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/partition.go:1376 +0x6a8
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.openPartition(0xc0000f8a40, 0x3b, 0xc0000f8a80, 0x39, 0xc0000eb3e0, 0xc0000f8834, 0x7, 0xc0000f8a80)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/partition.go:252 +0x1c6
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.openPartitions(0xc0000f85c0, 0x33, 0xc0000f86c0, 0x31, 0xc0000eb3e0, 0xc0000f8740, 0x3b, 0xc00011f7c8, 0x4520bf, 0x2d)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/table.go:452 +0x2a4
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.openTable(0xc00002b7a0, 0x2d, 0x12, 0xc0000eb3e0, 0x5, 0xc00002b7a0, 0x2d)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/table.go:116 +0x790
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.OpenStorage(0x7fff18f77e06, 0x28, 0x12, 0xab0fba, 0x35, 0xc00011fe78)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/storage.go:168 +0xcbd
main.main()
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmstorage/main.go:48 +0x1af
```

**To Reproduce**
- Initially data was collected when single-node had version `v1.33.0-0-g366693b9`.
- I've updated that single-node to `v1.34.9-0-g43375df9`, successfully started it and made `vmbackup` to s3 on it, the backup was successful and there is nothing special. (`backed up 855732 bytes in 9.089 seconds; deleted 0 bytes; server-side copied 0 bytes; uploaded 855732 bytes`)
- I've make `vmrestore` on vmstorage with version `v1.34.9-0-g43375df9` and it was successful too. (`restored 855732 bytes from backup in 3.583 seconds; deleted 0 bytes; downloaded 855732 bytes`)
- After that vmstorage crashes immediately after each startup attempt

I've also tried to remove whole `storageDataPath` directory and repeat restoring, but error still the same.

**Screenshots**
<details>
  <summary>Startup log</summary>

```

Started Victoria Metrics Storage.
2020-05-08T22:08:22.969Z        info        VictoriaMetrics/lib/logger/flag.go:11        build version: vmstorage-20200417-114202-tags-v1.34.9-cluster-0-g6764efde
2020-05-08T22:08:22.972Z        info        VictoriaMetrics/lib/logger/flag.go:12        command line flags
2020-05-08T22:08:22.972Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "bigMergeConcurrency" = "0"
2020-05-08T22:08:22.972Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "dedup.minScrapeInterval" = "0s"
2020-05-08T22:08:22.972Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "enableTCP6" = "false"
2020-05-08T22:08:22.973Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "envflag.enable" = "false"
2020-05-08T22:08:22.973Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "envflag.prefix" = ""
2020-05-08T22:08:22.973Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "fs.disableMmap" = "false"
2020-05-08T22:08:22.973Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "http.disableResponseCompression" = "false"
2020-05-08T22:08:22.973Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "http.maxGracefulShutdownDuration" = "7s"
2020-05-08T22:08:22.974Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "httpListenAddr" = ":8482"
2020-05-08T22:08:22.974Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "loggerFormat" = "default"
2020-05-08T22:08:22.974Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "loggerLevel" = "INFO"
2020-05-08T22:08:22.974Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "loggerOutput" = "stderr"
2020-05-08T22:08:22.975Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "memory.allowedPercent" = "70"
2020-05-08T22:08:22.975Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "precisionBits" = "64"
2020-05-08T22:08:22.975Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "retentionPeriod" = "18"
2020-05-08T22:08:22.975Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "rpc.disableCompression" = "false"
2020-05-08T22:08:22.975Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "search.maxTagKeys" = "secret"
2020-05-08T22:08:22.975Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "search.maxTagValues" = "100000"
2020-05-08T22:08:22.975Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "search.maxUniqueTimeseries" = "300000"
2020-05-08T22:08:22.975Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "smallMergeConcurrency" = "0"
2020-05-08T22:08:22.975Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "snapshotAuthKey" = "secret"
2020-05-08T22:08:22.975Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "storageDataPath" = "/media/vm-volume/var/lib/victoriametrics"
2020-05-08T22:08:22.975Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "version" = "false"
2020-05-08T22:08:22.975Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "vminsertAddr" = ":8400"
2020-05-08T22:08:22.975Z        info        VictoriaMetrics/lib/logger/flag.go:20        flag "vmselectAddr" = ":8401"
2020-05-08T22:08:22.975Z        info        VictoriaMetrics/app/vmstorage/main.go:46        opening storage at "/media/vm-volume/var/lib/victoriametrics" with retention period 18 months
2020-05-08T22:08:22.988Z        info        VictoriaMetrics/lib/memory/memory.go:35        limiting caches to 1463155097 bytes, leaving 627066471 bytes to the OS according to -memory.allowedPercent=70
2020-05-08T22:08:22.989Z        info        VictoriaMetrics/lib/storage/storage.go:680        loading MetricName->TSID cache from "/media/vm-volume/var/lib/victoriametrics/cache/metricName_tsid"...
2020-05-08T22:08:22.990Z        info        VictoriaMetrics/lib/storage/storage.go:685        loaded MetricName->TSID cache from "/media/vm-volume/var/lib/victoriametrics/cache/metricName_tsid" in 0.001 seconds; entrie
2020-05-08T22:08:22.993Z        info        VictoriaMetrics/lib/storage/storage.go:680        loading MetricID->TSID cache from "/media/vm-volume/var/lib/victoriametrics/cache/metricID_tsid"...
2020-05-08T22:08:22.994Z        info        VictoriaMetrics/lib/storage/storage.go:685        loaded MetricID->TSID cache from "/media/vm-volume/var/lib/victoriametrics/cache/metricID_tsid" in 0.001 seconds; entriesCou
2020-05-08T22:08:22.994Z        info        VictoriaMetrics/lib/storage/storage.go:680        loading MetricID->MetricName cache from "/media/vm-volume/var/lib/victoriametrics/cache/metricID_metricName"...
2020-05-08T22:08:22.996Z        info        VictoriaMetrics/lib/storage/storage.go:685        loaded MetricID->MetricName cache from "/media/vm-volume/var/lib/victoriametrics/cache/metricID_metricName" in 0.001 seconds
2020-05-08T22:08:22.996Z        info        VictoriaMetrics/lib/storage/storage.go:543        loading curr_hour_metric_ids from "/media/vm-volume/var/lib/victoriametrics/cache/curr_hour_metric_ids"...
2020-05-08T22:08:22.996Z        info        VictoriaMetrics/lib/storage/storage.go:546        nothing to load from "/media/vm-volume/var/lib/victoriametrics/cache/curr_hour_metric_ids"
2020-05-08T22:08:22.996Z        info        VictoriaMetrics/lib/storage/storage.go:543        loading prev_hour_metric_ids from "/media/vm-volume/var/lib/victoriametrics/cache/prev_hour_metric_ids"...
2020-05-08T22:08:22.997Z        info        VictoriaMetrics/lib/storage/storage.go:546        nothing to load from "/media/vm-volume/var/lib/victoriametrics/cache/prev_hour_metric_ids"
2020-05-08T22:08:23.001Z        info        VictoriaMetrics/lib/mergeset/table.go:166        opening table "/media/vm-volume/var/lib/victoriametrics/indexdb/15F8872989903790"...
2020-05-08T22:08:23.019Z        info        VictoriaMetrics/lib/mergeset/table.go:200        table "/media/vm-volume/var/lib/victoriametrics/indexdb/15F8872989903790" has been opened in 0.017 seconds; partsCount: 5; bl
2020-05-08T22:08:23.022Z        info        VictoriaMetrics/lib/mergeset/table.go:166        opening table "/media/vm-volume/var/lib/victoriametrics/indexdb/15F887298990378F"...
2020-05-08T22:08:23.038Z        info        VictoriaMetrics/lib/mergeset/table.go:200        table "/media/vm-volume/var/lib/victoriametrics/indexdb/15F887298990378F" has been opened in 0.017 seconds; partsCount: 3; bl
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x7c5b29]
goroutine 1 [running]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*indexBlockCache).MustClose(0x0, 0xc000052c00)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/part.go:206 +0x29
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*part).MustClose(0xc000236090)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/part.go:148 +0x8c
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.newPart(0xc0000d91d0, 0xc00008af80, 0x7c, 0x28e8e, 0xb56520, 0xc00007de90, 0xb565a0, 0xc0003577a0, 0xb565a0, 0xc0003577c0, ...)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/part.go:124 +0x359
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.openFilePart(0xc00008af80, 0x7c, 0xee16e0, 0xa919e5, 0x1)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/part.go:96 +0x7a8
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.openParts(0xc0000f8a40, 0x3b, 0xc0000f8a80, 0x39, 0xc0000f8a40, 0x3b, 0x0, 0x0, 0x0, 0x0, ...)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/partition.go:1376 +0x6a8
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.openPartition(0xc0000f8a40, 0x3b, 0xc0000f8a80, 0x39, 0xc0000eb130, 0xc0000f8834, 0x7, 0xc0000f8a80)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/partition.go:252 +0x1c6
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.openPartitions(0xc0000f85c0, 0x33, 0xc0000f86c0, 0x31, 0xc0000eb130, 0xc0000f8740, 0x3b, 0xc00011f7c8, 0x4520bf, 0x2d)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/table.go:452 +0x2a4
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.openTable(0xc00002af90, 0x2d, 0x12, 0xc0000eb130, 0x5, 0xc00002af90, 0x2d)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/table.go:116 +0x790
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.OpenStorage(0x7fff3031fe00, 0x28, 0x12, 0xab0fba, 0x35, 0xc00011fe78)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/storage.go:168 +0xcbd
main.main()
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmstorage/main.go:48 +0x1af
```

</details>

**Version**
The line returned when passing `--version` command line flag to binary. For example:
```
victoria-metrics-20200417-111608-tags-v1.34.9-0-g43375df9
vmbackup-20200417-111621-tags-v1.34.9-0-g43375df9
```

```
vmrestore-20200417-111627-tags-v1.34.9-0-g43375df9
vmstorage-20200417-114202-tags-v1.34.9-cluster-0-g6764efde
```

**Used command-line flags**
`-storageDataPath=/media/vm-volume/var/lib/victoriametrics -retentionPeriod=18 -memory.allowedPercent=70`

**Comment by valyala** [comment]
@nezed , thanks for the bug report!

The bug should be fixed in the commit bd5f4e0344d95a078984cf2cd21420fd4ed52d5c .

Note that single-node VictoriaMetrics has different storage data format comparing to cluster version. So it is impossible to copy storage data from single-node VictoriaMetrics to cluster version via backup / restore procedure - `vmstorage` should emit error about incompatible storage data when trying to start it on data from single-node VictoriaMetrics.

There are plans to implement efficient data migration tool from single-node VictoriaMetrics to cluster VictoriaMetrics. This tool must be included in [vmctl](https://github.com/VictoriaMetrics/vmctl). It is possible to use the following workaround for data migration from single-node VictoriaMetrics to cluster version:

1) To export all the data or the required part of the data from single-node VictoriaMetrics via [/api/v1/export](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/README.md#how-to-export-time-series). Note that the amounts of exported data could be big, so it is recommended compressing the data as described [in the docs](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/README.md#how-to-export-time-series). It is also recommended exporting different parts of data in parallel in order to speed up the process.
2) To import the exported data to VictoriaMetrics cluster via [/api/v1/import](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/README.md#how-to-import-time-series-data). The cluster Version has slightly different urls - see [these docs](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/cluster/README.md#url-format) for details.
3) To reset response caches on all the `vmselect` nodes in the cluster after the import is complete. This may be done either via calling `/internal/resetRollupResultCache` on every `vmselect` node or by deleting the directory pointed by `-cacheDataPath` command line flag on every `vmselect` node when it is stopped. Note that the cache is stored on graceful shutdown of `vmselect` node, so it must be deleted after `vmselect` node is stopped. See more details in [these docs](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/README.md#backfilling).

**Comment by valyala** [comment]
The bugfix for nil pointer dereference has been included in [v1.35.4](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). Closing this issue as fixed.
